### PR TITLE
feat(loans): amortisation engine and variable-rate loans

### LINKED ISSUE #3295: feat: Add variable interest rate loan amortization and tracking
## Problem Statement

Currently, Sure.am has limited support for variable interest rate loans. While the `rate_type` field exists in the loans table to support both "fixed" and "variable" loans, there is:

1. **No amortization schedule generation** - Loans cannot display detailed payment schedules showing principal/interest breakdown
2. **No variable rate tracking** - Variable interest rate changes over time are not tracked or stored
3. **No UI for viewing schedules** - Users cannot see payment-by-payment details
4. **No API endpoint** - External integrations cannot fetch amortization data

This is a significant gap for users managing variable-rate mortgages, ARMs, lines of credit, and other adjustable-rate loans.

## Current State

- ✅ `rate_type` field exists but only "fixed" is properly implemented
- ✅ `monthly_payment()` calculation only works for fixed rates
- ❌ No amortization schedule generation
- ❌ No variable rate change tracking
- ❌ No schedule rebuild when rates change
- ❌ No UI to view payment schedules
- ❌ No API endpoint for schedule data

### Backend
- [ ] AmortizationCalculator service for schedule generation
  - Support both fixed and variable rate calculations
  - Handle rate changes on specific dates
  - Accurate monthly compounding
  - Proper handling of final payments

- [ ] LoanAmortization model to store payment records
  - Track payment number, date, amount breakdown
  - Store effective interest rate per period
  - Enable efficient querying of payment history

- [ ] Loan model enhancements
  - Support for `variable_rate_schedule` (jsonb field for rate changes)
  - Support for `start_date` to track loan inception
  - `next_rate_change_date` optimization field
  - Automatic schedule rebuild when terms change
  - Methods for adding/updating variable rates

- [ ] Database schema updates
  - Migration for `loan_amortizations` table with proper indexes
  - Columns for `variable_rate_schedule` and `start_date` on loans table

### API
- [ ] REST endpoint: `GET /api/v1/loans/:id/amortization_schedule`
  - Return full payment schedule
  - Support pagination (limit/offset)
  - Include loan summary and rate details
  - OpenAPI documentation with rswag

### UI
- [ ] ViewComponent for amortization schedule display
  - Summary statistics dashboard
  - Interactive payment table
  - Variable rate change timeline
  - Payment breakdown visualization

- [ ] Stimulus controller for interactivity
  - CSV export functionality
  - Keyboard navigation
  - Rate change highlighting

- [ ] Page for viewing loan amortization
  - Print-friendly layout
  - Export to CSV/PDF options
  - Responsive design

### Testing
- [ ] Comprehensive test coverage
  - AmortizationCalculator for fixed/variable scenarios
  - LoanAmortization model validation
  - Loan model rate management
  - API endpoint tests
  - Edge cases (zero interest, zero principal, rate changes)

## Example Usage

```ruby
# Create a variable-rate loan
loan = Loan.create!(
  interest_rate: 5.0,
  term_months: 360,
  rate_type: "variable",
  start_date: Date.new(2024, 1, 1),
  variable_rate_schedule: {
    "2024-01-01" => 5.0,    # Initial rate
    "2025-01-01" => 5.5,    # Rate change 1
    "2026-01-01" => 6.0     # Rate change 2
  }
)

# Access amortization schedule
loan.amortization_schedule.each do |payment|
  puts "Payment #{payment.payment_number}: #{payment.payment_amount} " \
       "(#{payment.principal_payment} principal, #{payment.interest_payment} interest) @ #{payment.interest_rate}%"
end

# Add rate change
loan.add_variable_rate_change(Date.new(2027, 1, 1), 6.5)

# Get monthly payment
payment = loan.monthly_payment  # Returns Money object

# Access via API
GET /api/v1/loans/loan-id/amortization_schedule?limit=12
```

## Acceptance Criteria

- [x] Amortization schedules generated for both fixed and variable loans
- [x] Variable interest rate changes tracked and applied to calculations
- [x] Payment-by-payment details available (principal, interest, balance)
- [x] UI for viewing amortization schedules
- [x] API endpoint with full pagination support
- [x] OpenAPI documentation
- [x] Comprehensive test coverage (46+ tests)
- [x] i18n localization
- [x] Responsive and accessible design

## Technical Notes

- Uses BigDecimal for financial calculations (accurate to 4 decimal places)
- Maintains data invariant: ending_balance[n] = beginning_balance[n+1]
- Supports zero-interest loans
- Handles edge cases: zero principal, zero term
- Proper handling of floating-point rounding errors
- Atomic schedule rebuilds via database transactions

## Implementation Priority

**High** - This feature is essential for users with variable-rate loans and enhances the financial tracking capabilities significantly.

---

**Note:** This issue will be addressed by PR implementing the complete feature with backend, API, and UI components.

### PR DESCRIPTION
**Pull request by jaysbeekay**
Fixes #3295

Built on #2984. This is the first of two pull requests: #3474 (the loan balance chart with a payoff projection, for #3332) is stacked on this commit and can be reviewed after it, or folded in here if maintainers would rather review one change. It replaces #3296, which grew to 8,900 lines and was reviewed down; everything here is roughly 1,500 lines, and the two regressions that review found are carried as tests.

## What this does

Two slices, one commit:

**The calculation loop is replaced.** #2984 sizes one level payment up front and loops the term at a single `annual_rate`; there is no seam a rate change can enter through. The loop becomes `Loan::Simulator`, fed by callables, with `Loan::AmortizationMath` (the annuity formula and one period's split) and `Loan::SimulationResult`. The public API of `Loan::AmortizationSchedule` is unchanged: `payments`, `periodic_payment`, `total_interest`, `total_paid`, `payoff_date`, `payment_for` and the `Payment` Data type are exactly as reviewed, and **#2984's own test file passes byte-unmodified** against the new engine (`git diff` on `test/models/loan/amortization_schedule_test.rb` between the two is empty). That was the acceptance criterion for the swap.

**Variable-rate loans.** One migration adds `loans.variable_rate_schedule` (jsonb, effective date to annual percentage) and `loans.start_date`. `Loan::RateResolver` reads the recorded changes. The loan form gains a rate-change editor: plain array-named inputs, so it works without JavaScript, and a Stimulus controller adds and removes rows and hides the section for fixed loans. The Schedule tab re-amortises at each recorded change and labels the first figure "Opening Payment" when the repayment moves. The Overview tab shows the rate in force today, the original principal and the original payoff date. `amortizable?` widens to any loan with a non-blank rate type, so the Schedule tab appears for variable and adjustable loans, and for a provider's own vocabulary (`PlaidAccount::Liabilities::MortgageProcessor` writes Plaid's raw `interest_rate.type` straight through). A blank rate type is still not schedulable. `monthly_payment` stays nil for a variable loan: it has a schedule now, but not a single monthly payment, and quoting the one it opened with would present a stale figure as current.

A rate change dated before the loan's origination (its start date, or its first valuation when there is none) is rejected, and so is a future start date.

## Two rates, not one

A period is bounded by two dates, and on a variable loan they can sit either side of a change.

- **Interest** for `[previous payment date, this payment date)` accrues at the rate in force when the period **opened**.
- **Payment sizing** uses the rate in force **on** the payment date.

Reading one rate for both re-rates the month ending on the boundary, billing a rate that applied for none of it. This is also why #3296's second blocking finding is unreachable here: nothing counts days into payments (segments are real payment dates), so a segment spanning a 30-day month cannot round to zero payments and vanish. Pinned end-to-end on charged interest, because a skipped segment leaves a well-formed schedule and raises nothing.

## Deliberately not here

Daily accrual, offset accounts, extra repayments, scenarios, day-count conventions, a persisted schedule cache and its rebuild pipeline. Two expectations in #3295 are declined rather than omitted silently: monthly accrual applies a mid-period rate change from the **next** period (a property of monthly accrual, documented on `Loan::Simulator` where it happens; splitting periods at change dates is daily accrual), and there is no `Rails.cache` layer in this slice. Measured on an M-series laptop: a 360-month fixed schedule builds in 8 ms, a 360-month variable schedule with five changes in 63 ms, a 1,200-month schedule in 21 ms.

## Demo data

`Demo::Generator` gives the demo family's three loans real terms, so each has a schedule:

- **Home Mortgage:** adjustable, 320,000 at 6.25% over 30 years, with two recorded rate changes (5.5% two years in, 6.75% two years after that), so the demo shows a schedule re-amortising.
- **Car Loan:** fixed, 24,000 at 6.9% over 5 years.
- **Student Loan:** fixed, 42,000 at 5.5% over 10 years, with one extra payment a year ago.

Each loan opens at its principal and is paid from its own schedule, so its balance today is what the schedule says (less the student loan's extra payment). `test/models/demo/generator_test.rb` asserts the terms and the balances.

## Screenshots

From the demo family's Home Mortgage (adjustable, with two recorded rate changes). Taken on this pull request's previous revision (`33687b9b8`); the latest revision changes only a test, so every pictured screen is the same.

**Schedule tab.** The first card reads Opening Payment rather than Monthly Payment, because the repayment changes at each recorded rate change.

Schedule tab on the demo mortgage, light theme

Schedule tab on the demo mortgage, dark theme

**Overview tab.** The rate in force today, the original loan amount and the original payoff date.

Overview tab on the demo mortgage, light theme

Overview tab on the demo mortgage, dark theme

**Loan form.** The recorded rate changes, with the origination date.

Rate changes on the demo mortgage's edit form, light theme

Rate changes on the demo mortgage's edit form, dark theme

## Migration notes

One additive migration, `20260908145118_add_variable_rate_tracking_to_loans`:

- `loans.variable_rate_schedule`: `jsonb`, `default: {}`, `null: false`.
- `loans.start_date`: `date`, nullable.

Existing loans are not rewritten: they get an empty rate schedule and no start date, and behave as before; a loan without a start date takes its origination from its first valuation. Nothing to run after deploy.

## Blast radius

`amortizable?` is read by the account page's tabs, `AmortizationSchedule.for` and `Loan#monthly_payment`. The Schedule tab now appears on variable and adjustable loans; `monthly_payment` is unchanged. `amortizable?` also requires an account, which #2984's `rate_type == "fixed"` guard happened to short-circuit before. A term longer than 1,200 months is not amortizable rather than raising; `term_months` is also written by `PlaidAccount::Liabilities::StudentLoanProcessor` from provider dates, so a hard validation would turn one bad date into a failed sync.

`AccountableResource#update`, the edit action every account type shares, now runs the balance change and the attribute update in one transaction, so a rejected edit no longer persists half of itself. Every account type's controller test runs against it.

**Carried, unrelated to loans:** `Account#assign_default_owner` orders its candidates by `created_at, id`. Two admin fixtures share a timestamp, so the default owner depended on the query plan and flaked CI. It is one line and can be dropped if you would rather take it separately.

## Verification

All new behaviour is covered by tests that were observed to fail before the change they pin, including: interest accrues at the opening rate and the repayment resizes on the payment date; a change landing on a payment date does not re-rate the month before it; a one-month rate spike in a 30-day month is charged; a fixed loan's schedule is byte-identical to #2984's; a provider rate type outside the form's vocabulary still yields a schedule; invalid rate-change rows are rejected with a 422 and redisplayed even when nothing else on the loan changed; a rate change before origination is rejected; a fixed loan's retained rate changes survive a switch to variable and back; a half-filled row on the new-loan form re-renders the form rather than the error page; a future `start_date` is rejected; a rejected update does not persist the balance change submitted with it; the Overview tab shows the rate in force.

Locally, on this commit with #3474's on top, over `main` at `ca4a6b4a0`:

- The loan, account, account-type controller, account chart component, period picker and i18n tests: 375 runs, 0 failures (4 skipped: the i18n key checks `main` also skips).
- `test/system/loan_payoff_chart_test.rb` in headless Chrome: 6 runs, 0 failures.
- `bin/rubocop` on the changed Ruby files, `biome lint` on the changed controller and `bin/brakeman`: clean. No ERB has changed since the last full `erb_lint --lint-all`.
- Full `bin/rails test`: 9,050 runs, 0 failures, 1 error (`RecurringTransaction::PipelineTest`, whose own `PG.connect` fails auth on this machine on any branch).

GitHub checks on this commit, including the unit and system test jobs, are green.

## History

Squashed from [jaysbeekay/sure#109](https://github.com/jaysbeekay/sure/pull/109), where the full review history lives: three rounds of cubic, CodeRabbit and Codacy, and an independent review whose blocking, should-fix and minor findings are all addressed on the branch. The fork's plan for this delivery is [`docs/plans/loan-chart-upstream-delivery.md`](https://github.com/jaysbeekay/sure/blob/main/docs/plans/loan-chart-upstream-delivery.md).

### CODE REVIEW

File: app/models/loan/simulator.rb [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        # Resize only when the sizing rate actually moves. Recomputing every
+        # period would be arithmetically identical while the rate holds, but it
+        # would also silently absorb a payment the borrower is contracted to,
+        # which is what `:hold` exists to refuse.
+        if payment.nil? || (@payment_strategy == :reamortize && sizing_rate != previous_sizing_rate)
+          payment = AmortizationMath.level_payment(
+            balance: balance,
+            monthly_rate: sizing_rate,
+            remaining_payments: @payment_schedule.length - index,
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply app/models/loan/simulator.rb:117]: Real, and reproduced exactly: 500,000 over 24 months, 6% to 18% effective on the month-six payment date gave a level payment of 24,394.64 and a final payment of 19,156.02. A realistic move, 6% to 6.5% on a 360-month loan, left a final payment of 1,750.81 against 3,158.89, so this was not only an extreme-case problem.

Fixed in `0724c935` by the first of your two options. `AmortizationMath.level_payment` takes the interest the straddled period actually charged and sizes the one figure that covers it and then amortises the remainder over the periods left at the new rate (`P = (B + I) / (1 + a(n-1, r))`). When the period's interest is at the sizing rate this reduces to the plain annuity, so the plain formula is kept for that case and fixed-rate schedules stay bit-identical (the #2984 test file and the 2,997.75 pin still pass). The resize still lands on the payment the change is effective on, as documented.

Test, observed to fail first with `|24394.64 - 19156.02| (5238.62) to be <= 1.0`: `a resized payment stays level through the final settlement` in `test/models/loan/simulator_test.rb`, for a change on a payment date and a change mid-period; it also asserts the principal portions still sum to the balance.

File: app/models/loan.rb [hunk 2]
```diff
@@ -11,35 +11,184 @@ class Loan < ApplicationRecord
...
+    rate.nil? ? interest_rate : BigDecimal(rate.to_s)
+  end
+
+  # Rows the form submitted that could not be parsed. Kept so the save can be
+  # rejected and the form re-rendered with what the user actually typed --
+  # dropping them silently loses a typo'd row between submit and redisplay,
+  # and the user is never told which one went.
+  attr_reader :invalid_rate_changes
+
+  validate :rate_changes_must_be_parseable
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply app/models/loan.rb:112]: Real. `set_current_balance` writes a valuation and the account's cached balance, and the loan's validation ran after it. Fixed in `0724c935`: `AccountableResource#update` runs the balance change and the attribute update inside one `Account.transaction` and raises `ActiveRecord::Rollback` on either failure, then renders the 422 outside the transaction. The in-memory account keeps what the user typed, so the form redisplays it.

Test, observed to fail first (`Entry.count` changed by one and the balance moved on a 422): `a rejected update does not persist the balance change submitted with it` in `test/controllers/loans_controller_test.rb`. The depository, credit-card and property controller suites pass unchanged, since the concern is shared.

File: db/migrate/20260908145118_add_variable_rate_tracking_to_loans.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class AddVariableRateTrackingToLoans < ActiveRecord::Migration[7.2]
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply db/migrate/20260908145118_add_variable_rate_tracking_to_loans.rb]: Taken in `0724c935`: the migration now declares `ActiveRecord::Migration[8.1]`. For the record, the fork's CodeRabbit review had asked for the opposite, matching the 411 migrations at `[7.2]`; AGENTS.md line 13 is explicit that new migrations use the current version, so that rule wins here.

File: app/views/loans/_form.html.erb [hunk 4]
```diff
@@ -18,17 +20,77 @@
...
+          <%= t("loans.form.rate_changes_empty") %>
+        </p>
+
+        <template data-loan-rate-changes-target="template">
+          <%= render "loans/rate_change_row", effective_date: nil, rate: nil %>
+        </template>
+
+        <button type="button" class="text-sm text-primary underline"
+                data-action="loan-rate-changes#add"
+                <%= "disabled" if rate_changes_disabled %>>
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply app/views/loans/_form.html.erb]: Taken in `0724c935`: both controls render `DS::Button` (`variant: "ghost"`, `type: "button"`, `disabled:` passed through for a fixed loan, Stimulus `data-action` as before; Add carries the `plus` icon). The row partial doubles as the Stimulus clone template and still does. Test: `the rate-change controls are design-system buttons`, red before with no DS button span rendered; the existing tests that assert the controls are disabled server-side for a fixed loan still pass.

File: app/controllers/concerns/accountable_resource.rb [hunk 5]
```diff
@@ -56,6 +56,15 @@ def create
...
                 notice: t("accounts.create.success", type: accountable_type.name.underscore.humanize)
+  rescue ActiveRecord::RecordInvalid => e
+    # `create_and_sync` saves with `save!`. A validation the user can trip from
+    # the form (a half-filled loan rate change, say) must come back as the form
+    # with the error and what they typed, not as the generic 422 page. The
+    # transaction above has already rolled back; the record on the exception is
+    # the unsaved account with its nested accountable still attached.
+    @account = e.record
+    @error_message = e.record.errors.full_messages.join(", ")
+    render :new, status: :unprocessable_entity
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply app/controllers/concerns/accountable_resource.rb:78]: Taken in `0724c935`: the rescue calls `set_link_options` before rendering `new`, as the `new` action's before_action does. Test: `a rejected create with the method-select step still renders`, red before with `undefined method 'any?' for nil`.

**Comment by jaysbeekay** [comment]
All five review threads are addressed on the new head **`0724c935`** (force-pushed over `d30d2320`; this branch is the fork's reviewed branch squashed onto `main`, so each revision replaces the single commit rather than appending). The fork-side commit with the same changes and the observed-to-fail transcripts is jaysbeekay/sure@50f94cb8.

| Thread | Outcome |
| --- | --- |
| Codex: resized payment not level to maturity | **Real, fixed.** Reproduced at 5,238.62 short on the 24-month example and 1,408 short on a realistic 360-month move; sizing now accounts for the straddled period's accrued interest. Fixed-rate schedules bit-identical |
| Codex: balance change committed before a rejected validation | **Real, fixed.** One transaction, rolled back together |
| Codex: migration declares 7.2 | Taken, `[8.1]` per AGENTS.md |
| Codex: hand-built buttons | Taken, both controls are `DS::Button` |
| CodeRabbit: `@provider_configs` nil on the rescue render | Taken |

Each fix has a test that was observed to fail before it. On this exact tree: loan, provider, controller (loans, accounts, depositories), component and i18n suites 238 runs, 0 failures; the loan browser test 3 runs, 0 failures; rubocop, erb_lint, biome and brakeman clean. #3474 is rebuilt on this commit as `92210149`.

File: app/controllers/concerns/accountable_resource.rb [hunk 6]
```diff
@@ -56,25 +56,49 @@ def create
     return_path = safe_return_to(account_params[:return_to]) || session.delete(:return_to).presence || @account
     redirect_to return_path,
                 notice: t("accounts.create.success", type: accountable_type.name.underscore.humanize)
+  rescue ActiveRecord::RecordInvalid => e
+    # `create_and_sync` saves with `save!`. A validation the user can trip from
+    # the form (a half-filled loan rate change, say) must come back as the form
+    # with the error and what they typed, not as the generic 422 page. The
+    # transaction above has already rolled back; the record on the exception is
+    # the unsaved account with its nested accountable still attached.
+    @account = e.record
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply app/controllers/concerns/accountable_resource.rb]: Real: `OpeningBalanceManager#set_opening_balance` reaches `account.entries.create!`, and `lock_saved_attributes!` reaches `accountable.update!`, so `e.record` can be an `Entry` or the accountable. Fixed in `73a7368f`: the rescue keeps `e.record` when it is an `Account`, otherwise recovers the account through the record (`try(:account)`) and falls back to a fresh one as `new` builds it; the message falls back to `e.message` when the record carries no errors of its own. Test, observed to fail first with `undefined method 'loan' for an instance of Entry` out of the re-render: `a create that fails on the opening valuation still re-renders the form` in `test/controllers/loans_controller_test.rb`, stubbing `set_opening_balance` to raise `RecordInvalid` on an invalid `Entry`.

File: app/models/loan/simulator.rb [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        re_amortisation_events || ->(_from, _to) { [] }, :re_amortisation_events
+      )
+      @currency_precision = currency_precision
+      @payment_strategy = payment_strategy.to_sym
+      @settle_at_schedule_end = settle_at_schedule_end
+
+      raise ArgumentError, "payment schedule must not be empty" if @payment_schedule.empty?
+      if @payment_schedule.length > MAX_PERIODS
+        raise ArgumentError, "payment schedule exceeds #{MAX_PERIODS} periods"
+      end
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply app/models/loan/simulator.rb:76]: Declined, with the reasoning and the tests that pin it. `Loan#amortizable?` is false for a term over `Loan::Simulator::MAX_PERIODS` (`term_months.to_i <= Loan::Simulator::MAX_PERIODS` is one of its conditions), and the Schedule tab is rendered only when `amortizable?` is true (`UI::AccountPage#tabs`; a `tab=schedule` parameter for a non-amortizable loan falls back to the first tab). So `schedule.payments` in the view is never reached with a nil schedule: `AmortizationSchedule.for` returns nil and nothing renders it. Pinned by `a term longer than the simulator will walk is not amortizable, rather than raising` (`variable_rate_schedule_test.rb`, both sides of the boundary) and `hides the schedule tab when the loan cannot be amortized` (`loans_controller_test.rb`).

A `term_months` validation is deliberately not added: `PlaidAccount::Liabilities::StudentLoanProcessor` computes `term_months` from provider dates this app does not control and writes it with `update!`, so a hard validation would turn one bad provider date into a failed sync, and it would not help records already stored. The tolerant answer at `amortizable?` covers every record however it arrived. The reasoning is in the comment on that test.

File: app/models/loan/simulator.rb [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+          monthly_rate: accrual_rate,
+          currency_precision: @currency_precision,
+          final: final,
+          interest: interest
+        )
+
+        payments << {
+          payment_number: index + 1,
+          payment_date: payment_date,
+          interest_rate: BigDecimal(rate_on(payment_date).to_s),
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply app/models/loan/simulator.rb]: Taken in `73a7368f`, as you proposed: the row carries `interest_rate` (the rate the period opened with, which the interest column was computed with, so `beginning_balance * interest_rate / 12` reproduces the row) and `sizing_rate` (the rate the payment was sized at). They differ only on the row whose period straddles a change. Test, observed to fail first on that row's recomputation: `a re-amortisation event sizes the payment from its own effective date` in `test/models/loan/simulator_test.rb` now asserts both rates and the recomputed interest. `re_amortising?` reads `sizing_rate` (see the thread on #3474).

**Comment by jaysbeekay** [comment]
Second review round addressed on the new head **`73a7368f`** (force-pushed over `0724c935`; this branch is the fork's reviewed branch squashed onto `main`, so each revision replaces the single commit). Fork-side commit with the observed-to-fail transcripts: jaysbeekay/sure@2e680bb0.

| Thread | Outcome |
| --- | --- |
| CodeRabbit: `e.record` may not be an `Account` in the create rescue | **Real, fixed.** Recovered through the record or rebuilt as `new` does |
| CodeRabbit: terms over `MAX_PERIODS` reach the Schedule tab | **Declined with evidence** on the thread: `amortizable?` gates the tab; two existing tests pin it; a `term_months` validation would fail provider syncs |
| CodeRabbit: row's `interest_rate` is the sizing rate | **Taken.** Rows carry `interest_rate` (accrual) and `sizing_rate` |
| CodeRabbit (on #3474): `re_amortising?` answers on event presence | **Real, fixed** here, since the file is PR-1's: it now asks the run whether the sizing rate moves between adjacent rows |

On this exact tree: loan, provider, controller, accounts, depository, component and i18n suites 241 runs, 0 failures; the loan browser test 4 runs, 0 failures; rubocop, erb_lint, biome and brakeman clean. #3474 is rebuilt on this commit as `50960942`.

**Comment by jjmata** [comment]
Just a heads up that before we proceed with moving this to maintainer review, we should also add demo data to the feature, so that it can be tested / shown to users easily.  Look at `models/demo/generator.rb` 🙏

**Review by jjmata — CHANGES_REQUESTED** [review]
Please add demo data (a screenshot or two would be nice too!) before we proceed ... 🚀

**Comment by jaysbeekay** [comment]
Force-pushed a rebuild. The branch is now one commit, b1ffa940, on current main (bbc42af5). That replaces the merge of main from yesterday. Its only conflict was db/schema.rb, and the schema here matches how you resolved it.

@jjmata the demo data is in. Demo::Generator now gives the demo family's three loans real terms, so each one has a schedule: an adjustable mortgage (320,000 at 6.25% over 30 years, dropping to 5.5% two years in and going up to 6.75% two years after that), plus a fixed car loan and a fixed student loan. Each opens at its principal. I'll add screenshots shortly.

Squashed from jaysbeekay/sure#109 at 692d143d, which picks up these fork commits since the last push:

- jaysbeekay/sure@a4a4aa1b: the demo loans, and the two CodeRabbit test fixes (the schedule tab test no longer breaks once the clock passes 2027, and the half-filled rate change test checks the actual response)
- jaysbeekay/sure@352b9970: one parsed reading of the rate change column shared by the resolver and the form, min/max on the term input, and a clearer error when a schedule would be too long
- jaysbeekay/sure@692d143d: rate change rows now render as labelled form fields like the rest of the loan form

**Comment by jaysbeekay** [comment]
Screenshots from the demo data. This is the demo family's Home Mortgage, the adjustable one with two recorded rate changes.

Schedule tab. The first card says Opening Payment rather than Monthly Payment, because the repayment changes at each recorded rate change:

Schedule tab on the demo mortgage

The rate changes on the loan form:

Rate changes on the demo mortgage's edit form

**Comment by jaysbeekay** [comment]
Force-pushed another rebuild. The branch is now ceb369a5, on the same main (bbc42af5), squashed from jaysbeekay/sure#109 at 3e43e1ae. That adds two fixes that came out of the review on #3474:

- jaysbeekay/sure@d674315f: an account edit now locks its saved attributes inside the update transaction, so if the lock fails the balance change rolls back with everything else
- jaysbeekay/sure@3e43e1ae: changing any input the loan's schedule is built from (rate changes, rate, term, rate type, start date), or reloading the loan, drops the memoised schedule

File: test/controllers/loans_controller_test.rb [hunk 9]
```diff
@@ -101,12 +101,294 @@ class LoansControllerTest < ActionDispatch::IntegrationTest
...
+    get account_path(@account, tab: "schedule")
+
+    assert_response :success
+    assert_not_equal flat_body, response.body,
+      "recording a rate change must change what the schedule tab renders"
+    # A substring free of characters ERB escapes -- the full string contains an
+    # apostrophe and renders as &#39;.
+    assert_match "re-amortises at each recorded change", response.body
+    assert_match I18n.t("loans.tabs.schedule.opening_payment"), response.body,
+      "a re-amortising schedule must not label its first payment as THE monthly payment"
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply test/controllers/loans_controller_test.rb]: Already the case on this head: the test derives the change date from `@account.loan.origination_date >> 12`, with a comment saying why a fixed date would stop re-amortising once the calendar passed it. Nothing further to change.

**Comment by jaysbeekay** [comment]
Force-pushed a rebuild. The branch is now one commit, 6ac75898, on the same main (bbc42af5), squashed from jaysbeekay/sure#109 at d67f501d. That carries a review pass on the fork branch (jaysbeekay/sure@d67f501d):

- the rate-change validation message now has a real locale entry (`config/locales/models/loan/{en,de}.yml`) instead of an inline English default
- `Loan#current_variable_rate` delegates to `RateResolver#accrual_rate_for` rather than re-implementing the same lookup
- `rate_changes=` parses dates with `Date.iso8601`, as `variable_rates` reads them back, instead of the more lenient `Date.parse`
- the two hand-written `.form-field` blocks in the rate-change row are one `loans/_rate_change_field` partial
- `SimulationResult#total_cost`, which nothing read, is gone

The PR description now has a "Cross-cutting changes" section calling out the transactional `AccountableResource#update` and the default-owner tie-break, since both touch every account type.

**Comment by jaysbeekay** [comment]
Force-pushed a rebuild. The branch is now one commit, `33687b9b8`, on current `main` (`ca4a6b4a0`). It is squashed from jaysbeekay/sure#109 at `7535e12a7` and replaces `6ac75898e`.

What the rebuild picks up since the last push (all on the fork branch first, each with a test that was observed to fail):

- **Owner review of the loan form and Overview tab:**
  - the Overview amounts are named as loan amounts;
  - the original payoff date is shown on the Overview tab;
  - the demo loans pay from their own schedules.
- **Balance update failure:** a failed balance update now keeps the form values the user typed.
- **Variable rates:** a rate change dated before the loan's origination (its start date, or its first valuation when there is none) is rejected, with en and de messages. A related fix: the validation no longer caches a missing Account association before the Account is attached.
- **Final-pass review:**
  - a comment on when loan autosave loads the account;
  - pinned tests for the simulator's final-period settlement, at one period and at a zero rate;
  - a declined `term_months` model validation, because Plaid's student loan processor writes that field from provider dates.

**Conflicts with current `main`:** only the `db/schema.rb` version line. It keeps `main`'s newer `2026_09_13_000000`; PR-1's migration (`20260908145118`) is additive and its columns merge cleanly.

**Validation on this tree:** on the rebuilt stack (this commit with #3474's on top): the brief's validation set (loan models, every account type's controller, the account chart components, the period picker and i18n) 322 runs, 0 failures (4 skipped: the i18n key checks upstream `main` also skips); the loan chart system test in headless Chrome 4 runs, 0 failures; `bin/rubocop` (2,635 files), `erb_lint --lint-all`, `biome lint` and `bin/brakeman` clean; full `bin/rails test` 9,050 runs, 0 failures, 1 error (`RecurringTransaction::PipelineTest`, whose own `PG.connect` fails auth on this machine on any branch).

Still owed before this leaves draft: a German speaker's pass over the new strings in `config/locales/views/loans/de.yml`.

**Comment by jaysbeekay** [comment]
@jjmata this is ready for maintainer review. Your request for demo data and screenshots is now answered in the description itself:

- **Demo data:** `Demo::Generator` gives the demo family's three loans real terms, including an adjustable mortgage with two recorded rate changes (details under *Demo data*).
- **Screenshots:** light and dark captures of the Schedule tab, the Overview tab and the rate-change form, taken from the demo mortgage on this commit (`33687b9b8`).

Since your review the branch has been rebuilt on current `main` (`ca4a6b4a0`), every review thread is resolved, CI is green, and the description has *Migration notes* and *Verification* sections. #3474 is stacked on this commit.

File: test/models/account_test.rb [hunk 10]
```diff
@@ -34,6 +41,25 @@ class AccountTest < ActiveSupport::TestCase
...
+      family.accounts.create!(
+        name: "Unowned tie-break account #{i}",
+        balance: 0,
+        currency: "USD",
+        accountable: Depository.new
+      ).owner
+    end
+
+    assert_equal "admin", owners.first.role
+    assert_equal owners.first, owners.last
```
[the remark this answers is not in the document]
Comment (jaysbeekay) [reply test/models/account_test.rb]: Valid, fixed. The test now asserts the winner is the tied admin with the lowest `id` (`family.users.where(role: "admin").order(:id).first`), not only that two calls agree.

- Fork: jaysbeekay/sure@55980ea8a on #109. Here: `a9458da5d`, which replaces `33687b9b8`; the rest of the commit is unchanged.
- Observed to fail without the tie-breaker: with `:id` dropped from the three orderings in `Account#assign_default_owner`, a different tied admin wins and the assertion fails.

**Review by jjmata — APPROVED** [review]
